sdk/core/core/src/az_http_response_parser.c

```diff
@@ -0,0 +1,242 @@
+    AZ_RETURN_IF_FAILED(az_span_reader_expect_char(p_reader, AZ_LF));
+    self->kind = AZ_HTTP_RESPONSE_BODY;
+  } else {
+    self->kind = AZ_HTTP_RESPONSE_HEADER;
+  }
+  return AZ_OK;
+}
+
+/**
+ * Status line https://tools.ietf.org/html/rfc7230#section-3.1.2
+ * HTTP-version SP status-code SP reason-phrase CRLF
+ */
+AZ_NODISCARD az_result az_http_response_parser_read_status(
+    az_http_response_parser * const self,
+    az_http_response_status_line * const out) {
+  AZ_CONTRACT_ARG_NOT_NULL(self);
+  AZ_CONTRACT_ARG_NOT_NULL(out);
+
+  az_span_reader * const p_reader = &self->reader;
+
+  // HTTP-version = HTTP-name "/" DIGIT "." DIGIT
+  // https://tools.ietf.org/html/rfc7230#section-2.6
+  AZ_RETURN_IF_FAILED(az_span_reader_expect_span(p_reader, AZ_STR("HTTP/")));
+  AZ_RETURN_IF_FAILED(az_span_reader_expect_digit(p_reader, &out->major_version));
+  AZ_RETURN_IF_FAILED(az_span_reader_expect_char(p_reader, '.'));
+  AZ_RETURN_IF_FAILED(az_span_reader_expect_digit(p_reader, &out->minor_version));
+
+  // SP = " "
+  AZ_RETURN_IF_FAILED(az_span_reader_expect_char(p_reader, ' '));
+
+  // status-code = 3DIGIT
+  {
+    uint16_t status_code = 0;
+    for (int i = 3; i > 0;) {
+      --i;
+      uint8_t digit = 0;
+      AZ_RETURN_IF_FAILED(az_span_reader_expect_digit(p_reader, &digit));
+      status_code = status_code * 10 + digit;
+    }
+    out->status_code = status_code;
+  }
+
+  // SP
+  AZ_RETURN_IF_FAILED(az_span_reader_expect_char(p_reader, ' '));
+
+  size_t const begin = p_reader->i;
+
+  // reason-phrase = *(HTAB / SP / VCHAR / obs-text)
+  // HTAB = "\t"
+  // VCHAR or obs-text is %x21-FF,
+  while (az_is_reason_phrase_symbol(az_span_reader_current(p_reader))) {
+    az_span_reader_next(p_reader);
+  }
+
+  out->reason_phrase = az_span_sub(p_reader->span, begin, p_reader->i);
+
+  // CR LF
+  AZ_RETURN_IF_FAILED(az_span_reader_expect_span(p_reader, AZ_STR(AZ_CRLF)));
+
+  // set state.kind of the next HTTP response value.
+  AZ_RETURN_IF_FAILED(az_http_response_parser_set_kind(self));
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD bool az_is_http_whitespace(az_result_byte const c) {
+  switch (c) {
+    case ' ':
+    case '\t':
+      return true;
+  }
+  return false;
+}
+
+/**
+ * https://tools.ietf.org/html/rfc7230#section-3.2
+ */
+AZ_NODISCARD az_result az_http_response_parser_read_header(
+    az_http_response_parser * const self,
+    az_http_response_header * const out) {
+  AZ_CONTRACT_ARG_NOT_NULL(self);
+  AZ_CONTRACT_ARG_NOT_NULL(out);
+
+  az_span_reader * const p_reader = &self->reader;
+
+  // header-field   = field-name ":" OWS field-value OWS
+
+  // field-name     = token
+  {
+    size_t const field_name_begin = p_reader->i;
+
+    // https://tools.ietf.org/html/rfc7230#section-3.2.6
+    // token = 1*tchar
+    // tchar = "!" / "#" / "$" / "%" / "&" / "'" / "*" / "+" / "-" / "." / "^" /
+    //         "_" / "`" / "|" / "~" / DIGIT / ALPHA;
+    // any VCHAR,
+    //    except delimiters
+    while (true) {
```

Comment:
doing this parsing ourselves makes me very uneasy, unless we can confirm we're always talking to azure. (Note: just always using https does not do this, unless we validate specifically against azure's cert, AND we need to make sure there's no way for an attacker to control the http responses azure returns (for example by directing us to an attacker controlled azure endpoint).

Reply:
@JeffreyRichter any comment on this?

Reply:
@barcharcraz , are you proposing an alternate solution?
I'm not sure what you think the problem is. Are you concerned that the buffer doesn't have well-formed http response payload in it? If so, how would you expect we be resilient to that?